icu4c/source/i18n/number_simple.cpp

```diff
@@ -0,0 +1,255 @@
+        return;
+    }
+    if (fData == nullptr) {
+        status = U_INVALID_STATE_ERROR;
+        return;
+    }
+    fData->quantity.roundToMagnitude(position, roundingMode, status);
+}
+
+void SimpleNumber::setMinimumIntegerDigits(uint32_t position, UErrorCode& status) {
+    if (U_FAILURE(status)) {
+        return;
+    }
+    if (fData == nullptr) {
+        status = U_INVALID_STATE_ERROR;
+        return;
+    }
+    fData->quantity.setMinInteger(position);
+}
+
+void SimpleNumber::setMinimumFractionDigits(uint32_t position, UErrorCode& status) {
+    if (U_FAILURE(status)) {
+        return;
+    }
+    if (fData == nullptr) {
+        status = U_INVALID_STATE_ERROR;
+        return;
+    }
+    fData->quantity.setMinFraction(position);
+}
+
+void SimpleNumber::truncateStart(uint32_t position, UErrorCode& status) {
+    if (U_FAILURE(status)) {
+        return;
+    }
+    if (fData == nullptr) {
+        status = U_INVALID_STATE_ERROR;
+        return;
+    }
+    fData->quantity.applyMaxInteger(position);
+}
+
+void SimpleNumber::setSign(USimpleNumberSign sign, UErrorCode& status) {
+    if (U_FAILURE(status)) {
+        return;
+    }
+    if (fData == nullptr) {
+        status = U_INVALID_STATE_ERROR;
+        return;
+    }
+    fSign = sign;
+}
+
+
+void SimpleNumberFormatter::cleanup() {
+    delete fOwnedSymbols;
+    delete fMicros;
+    delete fPatternModifier;
+    fOwnedSymbols = nullptr;
+    fMicros = nullptr;
+    fPatternModifier = nullptr;
+}
+
+SimpleNumberFormatter SimpleNumberFormatter::forLocale(const icu::Locale &locale, UErrorCode &status) {
+    return SimpleNumberFormatter::forLocaleAndGroupingStrategy(locale, UNUM_GROUPING_AUTO, status);
+}
+
+SimpleNumberFormatter SimpleNumberFormatter::forLocaleAndGroupingStrategy(
+        const icu::Locale &locale,
+        UNumberGroupingStrategy groupingStrategy,
+        UErrorCode &status) {
+    SimpleNumberFormatter retval;
+    retval.fOwnedSymbols = new DecimalFormatSymbols(locale, status);
+    if (U_FAILURE(status)) {
+        return retval;
+    }
+    if (retval.fOwnedSymbols == nullptr) {
+        status = U_MEMORY_ALLOCATION_ERROR;
+        return retval;
+    }
+    retval.initialize(locale, *retval.fOwnedSymbols, groupingStrategy, status);
+    return retval;
+}
+
+
+SimpleNumberFormatter SimpleNumberFormatter::forLocaleAndSymbolsAndGroupingStrategy(
+        const icu::Locale &locale,
+        const DecimalFormatSymbols &symbols,
+        UNumberGroupingStrategy groupingStrategy,
+        UErrorCode &status) {
+    SimpleNumberFormatter retval;
+    retval.initialize(locale, symbols, groupingStrategy, status);
+    return retval;
+}
+
+
+void SimpleNumberFormatter::initialize(
+        const icu::Locale &locale,
+        const DecimalFormatSymbols &symbols,
+        UNumberGroupingStrategy groupingStrategy,
+        UErrorCode &status) {
+    if (U_FAILURE(status)) {
+        return;
+    }
+
+    fMicros = new SimpleMicroProps();
```

Comment:
Could these props be a value field, rather than a pointer + heap allocation?

Reply:
SimpleMicroProps is defined in an internal header. I would need to move it to a public header and mark it `@internal`. I would also need to move all of its value fields to a public header, although it looks like it contains only pre-existing public types as fields.